core/src/bundle_stage.rs

```diff
@@ -122,6 +122,80 @@ struct AllExecutionResults {
     pub post_balances: (TransactionBalances, TransactionTokenBalances),
 }
 
+struct BundleReservedSpace {
+    current_allocated_cost: u64,
+    initial_allocated_cost: u64,
+}
+
+impl BundleReservedSpace {
+    fn reset_reserved_cost(&mut self, working_bank: &Arc<Bank>) {
+        self.current_allocated_cost = self.initial_allocated_cost;
+
+        // Reserve space from cost tracker
+        QosService::update_or_remove_transaction_costs(
+            vec![TransactionCost::new_with_capacity(1)].iter(),
+            vec![Ok(())].iter(),
+            Some(&vec![CommitTransactionDetails::Committed {
+                compute_units: self.current_allocated_cost,
+            }]),
+            working_bank,
+        );
+        info!(
+            "Slot: {} - Update reserved space to {}. Block Cost {}",
+            working_bank.slot(),
+            self.current_allocated_cost,
+            working_bank
+                .read_cost_tracker()
+                // Using unwrap here is scary, but that's the way it's done in QoS
+                .unwrap()
+                .block_cost()
+        );
+    }
+
+    fn use_reserved_cost(&mut self, bundle_cost: u64, working_bank: &Arc<Bank>) {
+        // If there is enough reserved space left, deduct bundle cost from reserved space before adding
+        if bundle_cost <= self.current_allocated_cost {
+            self.current_allocated_cost -= bundle_cost;
+            let mut dummy_tx_cost = TransactionCost::new_with_capacity(1);
+            dummy_tx_cost.bpf_execution_cost = bundle_cost;
+            QosService::remove_transaction_costs(
+                vec![dummy_tx_cost].iter(),
+                vec![Ok(())].iter(),
+                working_bank,
+            );
+            info!(
+                "Bundle uses {bundle_cost} of reserved space. Remaining: {}",
+                self.current_allocated_cost
+            );
+        } else {
+            info!(
+                "Bundle Cost {bundle_cost} exceeds available reserved space {}",
+                self.current_allocated_cost
+            );
+        }
+    }
+
+    fn update_reserved_cost(&mut self, working_bank: &Arc<Bank>) {
+        let target_reserved_cost = self.initial_allocated_cost * working_bank.tick_height()
+            / working_bank.ticks_per_slot();
+        if self.current_allocated_cost > target_reserved_cost {
+            let mut dummy_tx_cost = TransactionCost::new_with_capacity(1);
+            dummy_tx_cost.bpf_execution_cost = self.current_allocated_cost - target_reserved_cost;
+            QosService::remove_transaction_costs(
```

Comment:
ditto here should we be calling `update_or_remove_transaction_costs` and have it hit the code path that calls into `update_transaction_costs`?

Reply:
hm I think I see what you're doing:
1. on new slot "reserve" max allowed cost for bundles
2. on receiving a new bundle you free up space for that bundle
3. then try and allocate it again thru `select_transactions_per_cost`

Reply:
still grokking but if I'm understanding correctly, don't we run the risk of banking stage eating up the freed space b/w steps 2 and 3?